README.md
# gaia-filter [![Build Status](https://secure.travis-ci.org/qualiancy/gaia-filter.png?branch=master)](https://travis-ci.org/qualiancy/gaia-filter)

> Array filtering inspired by MongoDB's query language.

## Installation

### Node.js

`gaia-filter` is available on [npm](http://npmjs.org).

    $ npm install gaia-filter

### Component

`gaia-filter` is available as a [component](https://github.com/component/component).

    $ component install qualiancy/gaia-filter

## Usage

### filter (query)

* **@param** _{Object}_ query 
* **@return** _{filter}_  

The primary export is a function that will memoize a query
so that it can be used to filter many arrays. Furthermore,
there are several methods to adjust the output of the result
set.

The filter query language is similiar MongoDB's query language
in that it allows for nested statements, deep object property
detection, and conditionals such as `$and`, `$or`, and `$nor`.

```js
var filter = require('gaia-filter');

var dataComplex = [
    { a: { b: 100 }
    , c: 'testC'
    , d:
      [ { e: 'world' } ]
   }
 , { a: { b: 50 }
   , c: 'testC'
   , d:
     [ { e: 'universe' }
     , { e: 'galaxy' } ]
   }
];

var query1 = filter({ 'a.b': { $and: [ { $gt: 75 }, { $lt: 125 } ] }});
  , query2 = filter({ 'a.b': { $gt: 25, $lt: 75 }, 'd[0].e': { $eq: 'universe' } });

var res1 = query1.subset(dataComplex)  // result array will have the first item
  , res2 = query1.subset(dataComplex); // result array will have the second item
```

Filter allows for some flexibility when composing queries by making assumptions
based on sane defaults.

When a comparator is omitted when comparing a value, it assumes `$eq`.

```js
// the following are the same.
filter({ hello: { $eq: 'universe' } });
filter({ hello: 'universe' });
```

When multiple statements are listed, it assumes `$and`.

```js
// the following are the same.
filter({ $and: [ { hello: { $gt: 42 } }, { hello: { $lt: 84 } } ] });
filter({ hello: { $and: [ { $gt: 42 }, { $lt: 84 } ] } });
filter({ hello: { $gt: 42, $lt: 84 } });
```


### .test (data)

* **@param** _{Mixed}_ data 
* **@return** _{Boolean}_  result

Test a single data point against the query. Will return a
boolean indicating whether data passes query criteria.

```js
var query = filter({ hello: { $eq: 'universe' }})
  , pass = query.test({ hello: 'universe' });

pass.should.be.true;
```


### .subset (data)

* **@param** _{Array}_ data 
* **@return** _{Array}_  result

Test an array of data points against the query. Will return
an array of all data points that pass the query criteria.

```js
var query = { $or : [ { 'hello': true }, { 'universe': true } ] }
  , q = filter(query)
  , result = q.subset([
      { hello: true }
    , { universe: false }
    , { hello: false, universe: true }
    , { hello: false, universe: false }
  ]);

result.should
  .be.an('array')
  .and.have.length(2)
  .and.deep.equal([
      { hello: true }
    , { hello: false, universe: true }
  ]);
```


### .pass (data)

* **@param** _{Array}_ data 
* **@return** _{Array}_  result

Test an array of data points against the query. Will return
an array of boolean in the order of the original array that
indicate if the data at that index passed the query criteria.

```js
var query = { $or : [ { 'hello': true }, { 'universe': true } ] }
  , q = filter(query)
  , result = q.pass([
        { hello: true }
      , { universe: false }
      , { hello: false, universe: true }
      , { hello: false, universe: false }
    ]);

result.should
  .be.an('array')
  .and.have.length(4)
  .and.deep.equal([
      true
    , false
    , true
    , false
  ]);
```


### .index (data)

* **@param** _{Array}_ data 
* **@return** _{Array}_  result

Test an array of data points against the query. Will return
an array of numbers indicating the indexes of the original
array that passed the query criteria.

```js
var query = { $or : [ { 'hello': true }, { 'universe': true } ] }
  , q = filter(query)
  , result = q.index([
        { hello: true }
      , { universe: false }
      , { hello: false, universe: true }
      , { hello: false, universe: false }
    ]);

result.should
  .be.an('array')
  .and.have.length(2)
  .and.deep.equal([ 0, 2 ]);
```


### .comparators

The following are the comparators implemented
for use as part of a query. They can also be used
for simple comparation, such as `filter.comparators.$eq(2,2)`,
which would return `true`.

##### $gt (a, b)

* **@param** _{Number}_ a 
* **@param** _{Number}_ b 
* **@return** _{Boolean}_  

Assert `a` is greater than `b`.


##### $gte (a, b)

* **@param** _{Number}_ a 
* **@param** _{Number}_ b 
* **@return** _{Boolean}_  

Assert `a` is greater than or equal to `b`.


##### $lt (a, b)

* **@param** _{Number}_ a 
* **@param** _{Number}_ b 
* **@return** _{Boolean}_  

Assert `a` is less than `b`.


##### $lte (a, b)

* **@param** _{Number}_ a 
* **@param** _{Number}_ b 
* **@return** _{Boolean}_  

Assert `a` is less than or equal to `b`.


##### $all (a, b)

* **@param** _{Array}_ a 
* **@param** _{Array}_ b 
* **@return** _{Boolean}_  

Assert `a` contains at least all items in `b`.


##### $exists (a, b)

* **@param** _{Mixed}_ a 
* **@param** _{Boolean}_ b 
* **@return** _{Boolean}_  

Assert truthiness of `a` equals `b`.


##### $mod (a, b)

* **@param** _{Number}_ a 
* **@param** _{Array}_ b 
* **@return** _{Boolean}_  

Assert `a` mod (`%`) `b[0]` equals `b[1]`.


##### $eq (a, b)

* **@param** _{Mixed}_ a 
* **@param** _{Mixed}_ b 
* **@return** _{Boolean}_  

Assert `a` equals (`===`) `b`.


##### $eq (a, b)

* **@param** _{Mixed}_ a 
* **@param** _{Mixed}_ b 
* **@return** _{Boolean}_  

Assert `a` does not equal (`!==`) `b`.


##### $in (a, b)

* **@param** _{Mixed}_ a 
* **@param** _{Array}_ b 
* **@return** _{Boolean}_  

Assert `a` is in `b` using `indexOf`.


##### $nin (a, b)

* **@param** _{Mixed}_ a 
* **@param** _{Array}_ b 
* **@return** _{Boolean}_  

Assert `a` is not in `b` using `indexOf`.


##### $size (a, b)

* **@param** _{Mixed}_ a 
* **@param** _{Number}_ b 
* **@return** _{Boolean}_  

Assert `a` has length of `b`. Returns
`false` if `a` does not have a length.
property.


##### $or (a)

* **@param** _{Array}_ a 
* **@return** _{Boolean}_  

Assert `a` has at least one truthy value.


##### $nor (a)

* **@param** _{Array}_ a 
* **@return** _{Boolean}_  

Assert `a` has zero truthy values.


##### $and (a)

* **@param** _{Array}_ a 
* **@return** _{Boolean}_  

Assert `a` has all truthy values.

## License

(The MIT License)

Copyright (c) 2012 Jake Luer <jake@qualiancy.com> (http://qualiancy.com)

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.


History.md

0.1.0 / 2012-11-18 
==================

  * typos
  * docs
  * tests
  * import liv
  * deps
  * makefile
  * "Initial commit"


test/input.js
describe('input', function () {
  describe('single comparator', function () {
    it('should return correct results', function () {
      var query = { $lt: 10 }
        , q = filter(query);
      q.test(8).should.be.true;
      q.test(11).should.be.false;
    });
  });

  describe('multiple comparators', function () {
    it('should return correct results', function () {
      var query = { $lt: 10, $gt: 5 }
        , q = filter(query);
      q.test(4).should.be.false;
      q.test(8).should.be.true;
      q.test(11).should.be.false;
    });
  });

  describe('nested comparators', function () {
    it('should return correct results', function () {
      var query = { $and: [ { $size: 3 }, { $all: [ 1, 2 ] } ] }
        , q = filter(query);
      q.test([0,1,2]).should.be.true;
      q.test([0,1,2,3]).should.be.false;
    });
  });
});


test/output.js
describe('output', function () {
  describe('.test(item)', function () {
    it('should return a boolean', function () {
      var query = { $or : [ { 'hello': true }, { 'universe': true } ] }
        , q = filter(query)
        , pass = q.test({ hello: true })
        , fail = q.test({ universe: false });

      pass.should
        .be.a('boolean')
        .and.equal(true);
      fail.should
        .be.a('boolean')
        .and.equal(false);
    });
  });

  describe('.subset(arr)', function () {
    it('should return an array of matched objects', function () {
      var query = { $or : [ { 'hello': true }, { 'universe': true } ] }
        , q = filter(query)
        , pass = q.subset([
              { hello: true }
            , { universe: false }
            , { hello: false, universe: true }
            , { hello: false, universe: false }
          ]);

      pass.should
        .be.an('array')
        .and.have.length(2)
        .and.deep.equal([
            { hello: true }
          , { hello: false, universe: true }
        ]);
    });
  });

  describe('.pass(arr)', function () {
    it('should return an array of boolean match results', function () {
      var query = { $or : [ { 'hello': true }, { 'universe': true } ] }
        , q = filter(query)
        , pass = q.pass([
              { hello: true }
            , { universe: false }
            , { hello: false, universe: true }
            , { hello: false, universe: false }
          ]);

      pass.should
        .be.an('array')
        .and.have.length(4)
        .and.deep.equal([
            true
          , false
          , true
          , false
        ]);
    });
  });

  describe('.index(arr)', function () {
    it('should return an array of indexes that match', function () {
      var query = { $or : [ { 'hello': true }, { 'universe': true } ] }
        , q = filter(query)
        , pass = q.index([
              { hello: true }
            , { universe: false }
            , { hello: false, universe: true }
            , { hello: false, universe: false }
          ]);

      pass.should
        .be.an('array')
        .and.have.length(2)
        .and.deep.equal([ 0, 2 ]);
    });
  });
});


test/assumptions.js
describe('non-comparator input assumptions', function () {
  describe('string', function () {
    it('should assume $eq', function () {
      var query = { 'hello': 'universe' }
        , q = filter(query);
      q.test({ hello: 'universe' }).should.be.true;
    });
  });

  describe('number', function () {
    it('should assume $eq', function () {
      var query = { 'hello': 42 }
        , q = filter(query);
      q.test({ hello: 42 }).should.be.true;
    });
  });

  describe('boolean', function () {
    it('should assume $eq', function () {
      var query = { 'hello': true }
        , q = filter(query);
      q.test({ hello: true }).should.be.true;
    });
  });

  describe('nested', function () {
    it('should assume $eq', function () {
      var query = { $or : [ { 'hello': true }, { 'universe': true } ] }
        , q = filter(query);
      q.test({ hello: true }).should.be.true;
      q.test({ universe: true }).should.be.true;
      q.test({ hello: false, universe: true }).should.be.true;
      q.test({ hello: false, universe: false }).should.be.false;
    });
  });

  describe('multiple nested', function () {
    it('should assume $and traversal', function () {
      var query = { 'test': 'hello', world: { $in: [ 'universe' ] } }
        , q = filter(query);
      q.test({ test: 'hello', world: 'universe' }).should.be.true;
      q.test({ test: 'hello', world: 'galaxy' }).should.be.false;
    });
  });
});


lib/filter.js
/*!
 * gaia-filter : MongoDB style array filtering
 * Copyright(c) 2011 Jake Luer <jake@alogicalparadox.com>
 * MIT Licensed
 */

/*!
 * Module dependancies
 */

var properties = require('tea-properties');

/*!
 * Constants
 */

var TRAVERSABLE = {
    $and: true
  , $or: true
  , $nor: true
};

/*!
 * Primary Exports
 */

module.exports = Filter;

/**
 * ### filter (query)
 *
 * The primary export is a function that will memoize a query
 * so that it can be used to filter many arrays. Furthermore,
 * there are several methods to adjust the output of the result
 * set.
 *
 * The filter query language is similiar MongoDB's query language
 * in that it allows for nested statements, deep object property
 * detection, and conditionals such as `$and`, `$or`, and `$nor`.
 *
 * ```js
 * var filter = require('gaia-filter');
 *
 * var dataComplex = [
 *     { a: { b: 100 }
 *     , c: 'testC'
 *     , d:
 *       [ { e: 'world' } ]
 *    }
 *  , { a: { b: 50 }
 *    , c: 'testC'
 *    , d:
 *      [ { e: 'universe' }
 *      , { e: 'galaxy' } ]
 *    }
 * ];
 *
 * var query1 = filter({ 'a.b': { $and: [ { $gt: 75 }, { $lt: 125 } ] }});
 *   , query2 = filter({ 'a.b': { $gt: 25, $lt: 75 }, 'd[0].e': { $eq: 'universe' } });
 *
 * var res1 = query1.subset(dataComplex)  // result array will have the first item
 *   , res2 = query1.subset(dataComplex); // result array will have the second item
 * ```
 *
 * Filter allows for some flexibility when composing queries by making assumptions
 * based on sane defaults.
 *
 * When a comparator is omitted when comparing a value, it assumes `$eq`.
 *
 * ```js
 * // the following are the same.
 * filter({ hello: { $eq: 'universe' } });
 * filter({ hello: 'universe' });
 * ```
 *
 * When multiple statements are listed, it assumes `$and`.
 *
 * ```js
 * // the following are the same.
 * filter({ $and: [ { hello: { $gt: 42 } }, { hello: { $lt: 84 } } ] });
 * filter({ hello: { $and: [ { $gt: 42 }, { $lt: 84 } ] } });
 * filter({ hello: { $gt: 42, $lt: 84 } });
 * ```
 *
 * @param {Object} query
 * @return filter
 * @api public
 */

function Filter (query) {
  if (!(this instanceof Filter)) {
    return new Filter(query);
  }

  this.query = query;
  this.stack = parseQuery(query);
}

/**
 * ### .test (data)
 *
 * Test a single data point against the query. Will return a
 * boolean indicating whether data passes query criteria.
 *
 * ```js
 * var query = filter({ hello: { $eq: 'universe' }})
 *   , pass = query.test({ hello: 'universe' });
 *
 * pass.should.be.true;
 * ```
 *
 * @param {Mixed} data
 * @return {Boolean} result
 */

Filter.prototype.test = function (data, opts) {
  return testFilter(data, this.stack);
};

/**
 * ### .subset (data)
 *
 * Test an array of data points against the query. Will return
 * an array of all data points that pass the query criteria.
 *
 * ```js
 * var query = { $or : [ { 'hello': true }, { 'universe': true } ] }
 *   , q = filter(query)
 *   , result = q.subset([
 *       { hello: true }
 *     , { universe: false }
 *     , { hello: false, universe: true }
 *     , { hello: false, universe: false }
 *   ]);
 *
 * result.should
 *   .be.an('array')
 *   .and.have.length(2)
 *   .and.deep.equal([
 *       { hello: true }
 *     , { hello: false, universe: true }
 *   ]);
 * ```
 *
 * @param {Array} data
 * @return {Array} result
 */

Filter.prototype.subset = function (data) {
  var res = []
    , datum;

  for (var di = 0; di < data.length; di++) {
    datum = data[di];
    if (testFilter(datum, this.stack)) res.push(datum);
  }

  return res;
};

/**
 * ### .pass (data)
 *
 * Test an array of data points against the query. Will return
 * an array of boolean in the order of the original array that
 * indicate if the data at that index passed the query criteria.
 *
 * ```js
 * var query = { $or : [ { 'hello': true }, { 'universe': true } ] }
 *   , q = filter(query)
 *   , result = q.pass([
 *         { hello: true }
 *       , { universe: false }
 *       , { hello: false, universe: true }
 *       , { hello: false, universe: false }
 *     ]);
 *
 * result.should
 *   .be.an('array')
 *   .and.have.length(4)
 *   .and.deep.equal([
 *       true
 *     , false
 *     , true
 *     , false
 *   ]);
 * ```
 *
 * @param {Array} data
 * @return {Array} result
 */

Filter.prototype.pass = function (data) {
  var res = []
    , datum;

  for (var di = 0; di < data.length; di++) {
    datum = data[di];
    res.push(testFilter(datum, this.stack));
  }

  return res;
};

/**
 * ### .index (data)
 *
 * Test an array of data points against the query. Will return
 * an array of numbers indicating the indexes of the original
 * array that passed the query criteria.
 *
 * ```js
 * var query = { $or : [ { 'hello': true }, { 'universe': true } ] }
 *   , q = filter(query)
 *   , result = q.index([
 *         { hello: true }
 *       , { universe: false }
 *       , { hello: false, universe: true }
 *       , { hello: false, universe: false }
 *     ]);
 *
 * result.should
 *   .be.an('array')
 *   .and.have.length(2)
 *   .and.deep.equal([ 0, 2 ]);
 * ```
 *
 * @param {Array} data
 * @return {Array} result
 */

Filter.prototype.index = function (data) {
  var res = []
    , datum;

  for (var di = 0; di < data.length; di++) {
    datum = data[di];
    if (testFilter(datum, this.stack)) res.push(di);
  }

  return res;
};

/**
 * ### .comparators
 *
 * The following are the comparators implemented
 * for use as part of a query. They can also be used
 * for simple comparation, such as `filter.comparators.$eq(2,2)`,
 * which would return `true`.
 */

Filter.comparators = {

    /**
     * ##### $gt (a, b)
     *
     * Assert `a` is greater than `b`.
     *
     * @param {Number} a
     * @param {Number} b
     * @return {Boolean}
     * @api public
     */

    $gt: function (a, b) {
      return a > b;
    }

    /**
     * ##### $gte (a, b)
     *
     * Assert `a` is greater than or equal to `b`.
     *
     * @param {Number} a
     * @param {Number} b
     * @return {Boolean}
     * @api public
     */

  , $gte: function (a, b) {
      return a >= b;
    }

    /**
     * ##### $lt (a, b)
     *
     * Assert `a` is less than `b`.
     *
     * @param {Number} a
     * @param {Number} b
     * @return {Boolean}
     * @api public
     */

  , $lt: function (a, b) {
      return a < b;
    }

    /**
     * ##### $lte (a, b)
     *
     * Assert `a` is less than or equal to `b`.
     *
     * @param {Number} a
     * @param {Number} b
     * @return {Boolean}
     * @api public
     */

  , $lte: function (a, b) {
      return a <= b;
    }

    /**
     * ##### $all (a, b)
     *
     * Assert `a` contains at least all items in `b`.
     *
     * @param {Array} a
     * @param {Array} b
     * @return {Boolean}
     * @api public
     */

  , $all: function (a, b) {
      for (var i = 0; i < b.length; i++) {
        if (!~a.indexOf(b[i])) return false;
      }
      return true;
    }

    /**
     * ##### $exists (a, b)
     *
     * Assert truthiness of `a` equals `b`.
     *
     * @param {Mixed} a
     * @param {Boolean} b
     * @return {Boolean}
     * @api public
     */

  , $exists: function (a, b) {
      return !!a == b;
    }

    /**
     * ##### $mod (a, b)
     *
     * Assert `a` mod (`%`) `b[0]` equals `b[1]`.
     *
     * @param {Number} a
     * @param {Array} b
     * @return {Boolean}
     * @api public
     */

  , $mod: function (a, b) {
      return a % b[0] == b[1];
    }

    /**
     * ##### $eq (a, b)
     *
     * Assert `a` equals (`===`) `b`.
     *
     * @param {Mixed} a
     * @param {Mixed} b
     * @return {Boolean}
     * @api public
     */

  , $eq: function (a, b) {
      return a === b;
    }

    /**
     * ##### $eq (a, b)
     *
     * Assert `a` does not equal (`!==`) `b`.
     *
     * @param {Mixed} a
     * @param {Mixed} b
     * @return {Boolean}
     * @api public
     */

  , $ne: function (a, b) {
      return a !== b;
    }

    /**
     * ##### $in (a, b)
     *
     * Assert `a` is in `b` using `indexOf`.
     *
     * @param {Mixed} a
     * @param {Array} b
     * @return {Boolean}
     * @api public
     */

  , $in: function (a, b) {
      return ~b.indexOf(a) ? true : false;
    }

    /**
     * ##### $nin (a, b)
     *
     * Assert `a` is not in `b` using `indexOf`.
     *
     * @param {Mixed} a
     * @param {Array} b
     * @return {Boolean}
     * @api public
     */

  , $nin: function (a, b) {
      return ~b.indexOf(a) ? false : true;
    }

    /**
     * ##### $size (a, b)
     *
     * Assert `a` has length of `b`. Returns
     * `false` if `a` does not have a length.
     * property.
     *
     * @param {Mixed} a
     * @param {Number} b
     * @return {Boolean}
     * @api public
     */

  , $size: function (a, b) {
      return (a && a.length && b) ? a.length == b : false;
    }

    /**
     * ##### $or (a)
     *
     * Assert `a` has at least one truthy value.
     *
     * @param {Array} a
     * @return {Boolean}
     * @api public
     */

  , $or: function (a) {
      var res = false;
      for (var i = 0; i < a.length; i++) {
        var fn = a[i];
        if (fn) res = true;
      }
      return res;
    }

    /**
     * ##### $nor (a)
     *
     * Assert `a` has zero truthy values.
     *
     * @param {Array} a
     * @return {Boolean}
     * @api public
     */

  , $nor: function (a) {
      var res = true;
      for (var i = 0; i < a.length; i++) {
        var fn = a[i];
        if (fn) res = false;
      }
      return res;
    }

    /**
     * ##### $and (a)
     *
     * Assert `a` has all truthy values.
     *
     * @param {Array} a
     * @return {Boolean}
     * @api public
     */

  , $and: function (a) {
      var res = true;
      for (var i = 0; i < a.length; i++) {
        var fn = a[i];
        if (!fn) res = false;
      }
      return res;
    }
};

/*!
 * Given the query input, create a re-usable definition
 * for how to test data again the query.
 *
 * @param {Object} query
 * @returns {Array} stack to be used with `Filtr.prototype.test`
 */

function parseQuery (query) {
  var stack = []
    , params, qry;

  for (var cmd in query) {
    params = query[cmd];
    qry = {};

    if (cmd[0] == '$') {
      qry.test = parseFilter(query);
    } else if ('string' == typeof params
    || 'number' == typeof params
    || 'boolean' == typeof params) {
      qry.test = parseFilter({ $eq: params });
      qry.path = cmd;
    } else {
      qry.test = parseFilter(params);
      qry.path = cmd;
    }

    stack.push(qry);
  }

  return stack;
};

/*!
 * Given that the root object passed is a comparator definition,
 * return a consumable test definition.
 *
 * @param {Object} query
 * @returns {Array} stack for use as input with `testFilter`
 */

function parseFilter (query) {
  var stack = [];

  for (var test in query) {
    var fn = Filter.comparators[test]
      , params = query[test]
      , traverse = false
      , st = []
      , p, nq;

    if (TRAVERSABLE[test]) {
      traverse = true;
      for (var i = 0; i < params.length; i++) {
        p = params[i];
        if ('string' == typeof p
        ||  'number' == typeof p
        ||  'boolean' == typeof p) {
          traverse = false;
        } else {
          nq = parseQuery(p);
        }
        st.push(nq);
      }
    }

    stack.push({
        fn: fn
      , params: traverse ? st : params
      , traverse: traverse
    });
  }

  return stack;
};

/*!
 * Given a well-formed stack from `parseFilter`, test
 * a given value again the stack.
 *
 * As the value is passed to a comparator, if that comparator
 * cannot interpret the value, false will be return. IE $gt: 'hello'
 *
 * @param {Object} value for consumption by comparator test
 * @param {Array} stack from `parseFilter`
 * @return {Boolean} result
 * @api private
 */

function testFilter (val, stack) {
  var pass = true;

  for (var si = 0, sl = stack.length; si < sl; si++) {
    var filter = stack[si]
      , el = filter.path
        ? properties.get(val, filter.path)
        : val
      , res = _testFilter(el, filter.test);

    if (!res) pass = false;
  }

  return pass;
};

/*!
 * Proxy the testing to the comparator or back to `testFilter`
 * if we traversing.
 *
 * @param {Mixed} value to test
 * @param {Array} stack to iterate
 * @return {Boolean} result
 * @api private
 */

function _testFilter (val, stack) {
  var res = true;

  for (var i = 0; i < stack.length; i++) {
    var test = stack[i]
      , params = test.params
      , p = [];

    if (test.traverse) {
      for (var ii = 0; ii < params.length; ii++) {
        p.push(testFilter(val, params[ii]));
      }

      params = p;
    }

    if (test.fn.length === 1 && !test.fn(params)) res = false;
    else if (test.fn.length > 1 && !test.fn(val, params)) res = false;
  }

  return res;
};
